### PR TITLE
Added ANSI 256 terminal color support

### PR DESCRIPTION
**Pull request by lersi**
Hello,
Not all terminals support ANSI FULL-COLOUR (RGB), in example MacOS default terminal app.  
So I have added support for the older ANSI standard of 256 colors.  

For the constant colors, I hand picked the closest match in ANSI 256 colors, for the functions that receive an RGB color parameter, I used an LGPL3 library that converts RGB tp ANSI256 in a very fast (and less accurate) way.  
The down side is that we cannot include this library in release builds, because it will violate the LGPL3 license.
As long as we do not distribute binaries with LGPL3 component, we do not have to license this code base under LGPL3.  
I have farther explained this issue in the source files, and added disclaimers where needed.

### CODE REVIEW

**Comment by DangerousPrototypes** [comment]
Wow, this is an impressive contribution. I'm curious - is the library developer willing to release it dual license or whatever under MIT or BSD? It's dated 2018, so maybe 6 years later they won't mind? Or maybe it was just a off the cuff license choice?

**Comment by lersi** [comment]
Hello Ian,
Thank you very mach for the compliment!  

Contacting the developer is a great idea, I will try to contact her and see where it goes.  
I think that it is best to keep the RGB feature because the colors look much better.
In addition, 3rd party terminals do support this feature and the command `screen` would probably support it on the next major release (the feature is already supported on the master branch).

In order to properly support both modes, some major redesign and abstraction will need to happen, because every function that receives an RGB value will also need an ansi 256 matching function, or we will need to decide on a set of colors to work with, and each color will get it's RGB value and ANSI256 value.
The solution in the `ansi_colours` project lets us to keep the existing API and simply convert the RGB values behind the scenes.

**Comment by pfeerick** [comment]
> The down side is that we cannot include this library in release builds, because it will violate the LGPL3 license.

I'm not sure that is actually correct. Did you make any changes to the library? As long as you 1) don't make any changes 2) don't directly copy bits of the library into BP code, the LGPL should not apply to the BP project. Including the library via submodule may make this even more obvious that this is the case (a "work that uses the library"), but will probably confuse more people than it's worth. 

> **A program that contains no derivative of any portion of the Library**, but is designed to work with the Library by being compiled or linked with it, **is called a "work that uses the Library"**. Such a work, in isolation, is not a derivative work of the Library, and **therefore falls outside the scope of this License**.

Also, the [requirements for a "Combined work"](https://www.gnu.org/licenses/lgpl-3.0.en.html) are easily met. 

- a - notice in the readme that the library is being used, and is covered by AGPLv3
- b - copy of the GPL and AGPL licenses in the repo... they don't say where, so I would suggest in the library folder would be sufficient
- c - don't think this applies... if it does, just add the note next to existing mention of MIT license
- d - the first option (0) is already met since the BP project source is available to build for yourself
- e - nothing more than link to the library github repo is required since it is publicly avaialble ... i.e. could be satisfied as part of a) via a url link.

**Comment by lersi** [comment]
@pfeerick 
From my understanding we do not comply with section 4.d, under option 0, we have to make the entire code base under LGPL3 license, because the entire codebase interacts either directly with the library or with code that interacts with library.
because this is an embedded project and the final binary is a complete firmware, we cannot use option 1.

**Comment by pfeerick** [comment]
> we have to make the entire code base under LGPL3 license

While I'm somewhat converstant in legalase, I'm not a copyright lawyer. But, that doesn't match with the stated reason for the LGPL to even exist, namely "using the Lesser GPL permits use of the library in proprietary programs"[^1] (i.e. closed source). 

andd 

> The Lesser GPL was designed to fill the middle ground between these cases, allowing proprietary software developers to use the covered library, but providing a weak copyleft that gives users freedom regarding the **library code itself**.[^2]

Note the focus being on the library code itself... the goal has always to been to project the openess of the library, without burdening/locking out others. So, given 

> A “Combined Work” is a work produced by combining or linking an Application with the Library. 

defined the BP firmware a combined work for the purposes of the license, and 

> 4. Combined Works.
You may convey a Combined Work **under terms of your choice that**, taken together, effectively **do not restrict modification of the portions of the Library contained in the Combined Work** and reverse engineering for debugging such modifications, if you also do each of the following

only requires you do not restrict ability to modify the portion of the library contained, since in accordance with 4d0 you can recompile the entire code yourself, and that you can distribute under terms of your choice, I don't see it as saying that at all. 

Regardless, getting the original LGPL licence to be changed would be nice. However I don't see that being an issue, but won't harp on it any further. :rofl: 

[^1]: https://www.gnu.org/licenses/why-not-lgpl.html
[^2]: https://www.gnu.org/licenses/license-recommendations.html

File: docs/licenses.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+Forcing the end user to compile git himself (not an easy task) and this will occur in any GNU based linux distribution which is 
+insane. In contrast using Apple's propriety software gives the user better experience because there are no GNU shenanigans.  
+
+For the developer the rapist conditions that the GNU license imposes leaves the developer with 3 unattractive options:
+The first option is to ignore any GPL protected code, thus limiting his available options to incorporate from the open source 
+community. So his freedom is taken by limiting the available options which are in the firstplace shared so the community/humanity
+can enjoy the code/solution that have already been written.  
+
+The second option that the devloper has is to surrender to GPL's license thus making his entire code protected by GPL's license
+and now he is raping his users to use GPL's license.
```
Comment (mina86) [inline docs/licenses.md:54]: If you’re developing free software why is that even a problem?

If you’re not, than I don’t give a fuck about your right to develop closed source code.

**Comment by mina86** [comment]
IANAL but if this project is free software than 4d0 is quite trivially met.  All you need to do is point at this repository each time you distribute firmware built with `ansi_colours` included.  Technically, you may need to license the code under LGPLv3 but that only means that your code becomes dual-licensed and if someone doesn’t like LGPL they can stick to MIT and not include `ansi_colours`.  And this is a technicality I certainly don’t care in practice.

And if this project is not free software than a) you don’t get my sympathy and b) you can distribute the firmware as two object files one with all the closed source and the other with `ansi_colours`.  You may loose on some optimisations that way (though that may be addressed with LTO) but again, if you’re working on closed software you don’t get my sympathy.

**Comment by lersi** [comment]
Thank you Michal for the clarification, this project is completely open source, even the hardware design is open source.

@DangerousPrototypes do you wan to dual license this project?

**Comment by DangerousPrototypes** [comment]
I'm keen to be open as possible. The previous Bus Pirate stuff was all CC0/public domain, I use MIT now because I've hoovered up so much of the PICO SDK/examples and they're all MIT/BSD.

From a personal perspective, I'd prefer not to dual license the project. Not because I have anything against GPL, but because it's a long term additional commitment of hassle making sure we're always compliant. I also want to avoid license wars/trolls who aren't willing to understand the nuance of discussions and compromises like we're discussing here.

Did the library author reply by any chance? I know I've had very little success contacting people via github (I'm hard to contact via github too, but you'll always find me in our forum), but a more permissive license would be the ideal solution.

**Comment by lersi** [comment]
The library author is @mina86 and you can see her comments here

**Comment by pfeerick** [comment]
> The option 0, that we must take in order comply with the license, states the following:
> 
> > Convey the Minimal Corresponding Source under the terms of this License, and the Corresponding Application Code in a form suitable for, and under terms that permit, the user to recombine or relink the Application with a modified version of the Linked Version to produce a modified Combined Work, in the manner specified by section 6 of the GNU GPL for conveying Corresponding Source.
> 
> The section states `Convey the Minimal Corresponding Source under the terms of this License`, in our case it's the entire bus pirate 5 firmware code, since we don't have any dynamic linking mechanism.

Incorrect. Read the license definitions again. Bus Pirate is the Application - it is _not_ based on the library code -  and Minimal Corresponding Source  **excludes** the Application source. Minimum Corresponding source is the library code, with it's license intact (hence why it must be under terms of 'this license'). Corresponding Application Code includes the BP source code, but merely under terms that permit it being used. 

> I think that their general goal is that user will be free to change the libraries code and relink it with the binary, so any thing that using this library without a dynamic link, must release his source code under the same terms.

That would make 4d0 pointless, especially when considering "recombine". It may help to read the LGPLv2.1 and corresponding support documentation to fully understand what the LGPL aimed to do. Some of the meaning and context seems to have been lost in the v3 license. It was quite clear that a LGPL licensed library could be statically compiled into closed source (let alone permissively licensed code like under the MIT/X11 license) and be distributed, without being forced to use the LGPL license. There was even explicit coverage over how compilers in optimising code and thus actually incorporating library code into the application code was not a breach of the license, but simply a technological function of the compiler. The "and not on the Linked Version" at the end of the Minimal Corresponding Source definition somewhat alludes to that still.

**Comment by lersi** [comment]
Alright, I accept your arguments, I will rephrase the licensing and change the build mechanism to OPT-OUT.

Thank you very much @pfeerick for the detailed explanation, I am relatively new to open source licenses and this was the first time that I needed to look deeply into it.

**Comment by pfeerick** [comment]
It's been an ... Interesting ... rabbit hole for me... I am just glad I
finally found something in "plain English" that verified my interpretation
appears to be correct, rather than "it's just the vibe of the thing, your
honor, it's just the vibe" :rofl:

So thank you for the conundrum ;)

**Comment by lersi** [comment]
We just need to publish a zip file containing the licensing.md, MIT license, LGPL3 license and the firmware

**Comment by lersi** [comment]
Self reminder: 
On the way document other non MIT licenses used in this project

**Comment by lersi** [comment]
I have added also a contribution guideline, and documented all the copyright notices that I have found in this repo.
In addition, I rephrased the licensing explanation, and change the build system to opt-out (plus moved the cmake files to a new folder called `cmake`.  

@pfeerick please check if my explanation suites you

I still need to document the BSD3 code

**Review by pfeerick — COMMENTED** [review]
Mostly just some typos, and re-wording... use or not as you please ;)

I'm not really sure `docs/licenses.md` is needed at all... all that should really be needed is mention/attribution in the readme, since the license texts are publicly available (but including them in the repo does not hurt). 

As an example of how [another OSS project does this](https://github.com/arduino/arduino-ide):

> ## License
> 
> The code contained in this repository and the executable distributions are licensed under the terms of the GNU AGPLv3. The executable distributions contain third-party code licensed under other compatible licenses such as GPLv2, MIT and BSD-3. If you have questions about licensing please contact us at [license@arduino.cc](mailto:license@arduino.cc).

I did look for some others, but most of the ones I'm readily familiar with don't need this distinction so don't mention anything at all.

File: readme.md [hunk 2]
```diff
@@ -32,4 +32,44 @@ VT100 terminal emulation supports color and a live statusbar view of the voltage
...
+    `cmake -S . -B build`  
+    you may want to add the flag `-DPICO_SDK_PATH=/path/to/pico-sdk` if you want to use pico-sdk that is in your local path.
+2. project build  
+    `cmake --build ./build --target bus_pirate5_rev10`  
+    you may set `bus_pirate5_rev10` to `bus_pirate5_rev8` if the have the development version.
+
+### patching pico SDK
+By default, pico-sdk assumes that we have 2MB of flash available, and there is no way to configure different flash size.  
+In reality, buspirate5 has 16MB of flash, if we want to fully utilize the available flash,
+we need to do the following modifications (currently it is not needed because buspirate5 firmware size is less than 2MB).
```
[the remark this answers is not in the document]
Comment (lersi) [reply readme.md:51]: Note that I use the markdown feature of splitting a line into multiple lines, so the file is easy to edit.
The end result is the same

**Comment by lersi** [comment]
Thank you @pfeerick for the typos and going over my code!!

I will document now the usage of BSD 3 components

**Comment by lersi** [comment]
Ready for merge!

File: docs/licenses.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#### [ansi_colours](https://github.com/mina86/ansi_colours)
+##### Files that are being used from the project
+* `ansi256.c`
+* `ansi_colours.h`
+##### Source Files that use this library
+* `ui/ui_term.c`
+##### copyright notice
+Copyright 2018 by Michał Nazarewicz <mina86@mina86.com>
+
+### License Complience explained
```
Comment (pfeerick) [inline docs/licenses.md:48]:
```suggestion
### License Compliance explained
```

File: readme.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (pfeerick) [reply readme.md:59]: Fair enough... I prefer letting my editor do the word wrap due to using different screen sizes at different times, but that is entirely personal preference. Heading is still missing a capital, and that one was somewhat of a re-write, not just a whitespace change.

**Review by pfeerick — COMMENTED** [review]
Just a typo and a formatting nitpick left, then I'll leave you in peace :hand_over_mouth:

File: CMakeLists.txt [hunk 4]
```diff
@@ -230,4 +230,17 @@ endforeach()
 
 string(APPEND CMAKE_EXE_LINKER_FLAGS "-Wl,--print-memory-usage")
 
+
+# by default this project uses LGPL3 libraries, if you do not want to comply
+# with LGPL3 license, please set the following variable to FALSE, otherwise
+# you are oblicated to distribute the resulted binaries under LGPL3 license
```
[the remark this answers is not in the document]
Comment (pfeerick) [reply CMakeLists.txt:253]: Did you want to change this, or just correct the typo?

**Comment by lersi** [comment]
@DangerousPrototypes what else should I fix in order to merge this pull request?

**Comment by DangerousPrototypes** [comment]
We'll see how this change goes :) Thank you so much for your efforts on this.

**Pull request merged**